solutions/usaco-1206.mdx

```diff
@@ -0,0 +1,107 @@
+---
+id: usaco-1206
+source: USACO Silver 2022 February
+title: Redistributing Gifts
+author: Chongtian Ma
+---
+
+[Official Analysis](http://www.usaco.org/current/data/sol_prob1_silver_feb22.html)
+
+## Explanation
+
+Similar to the official analysis, we created a directed graph with edges from $i$ to $j$ if cow $i$ prefers gift $j$ more compared to his own. 
+If cow $i$ is able to receive a gift where it has an outgoing edge, then the solution will always be more optimal. 
+Remember to also add an edge between cow $i$ and $i$ in the case where it cannot get a more preferred gift.
+
+Then, I separated the graph into strongly connected components. If $i$ is in a strongly connected component, this means that cow $i$ can receive
+all gifts in the strongly connected component through a series of reassignments and is guaranteed to have one or more preferred gift (including itself) based 
+on SCC definition. Obviously you only want to assign cow $i$ with a more preferred gift.
+## Implementation
+
+I used Kosaraju's algorithm to find all strongly connected components in the graph. At the end, I just went down cow $i$'s preference list and assigned the first gift belonging in the SCC.
+Time complexity is $O(M)$ where $M$ is the number of edges as mentioned at the bottom of the official analysis.
+
+<LanguageSection>
+<CPPSection>
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+#define all(x) x.begin(), x.end()
+const int N = 505;
+
+// for kosaraju purposes
+vector<int> graph[N], revgraph[N]; // revgraph is the reversed graph
+vector<bool> vis(N), vis2(N);
+vector<int> path; 
+
+// stores all nodes in a SCC
+set<int> comp;
+
+void dfs(int node){
+    vis[node] = true;
+    for(int i: graph[node]){
+        if(!vis[i]){
+            dfs(i);
+        }
+    }
+    path.push_back(node);
+}
+
+void dfs2(int node){
+    vis2[node] = true;
+    for(int i: revgraph[node]){
+        if(!vis2[i]){
+            dfs2(i);
+        }
+    }
+    comp.insert(node);
+}
+signed main() {
+    ios::sync_with_stdio(0);
+    cin.tie(0);
+    int n; cin>>n;
+    for(int i = 1; i <= n; i++){
+        // NOTE that graph array serves as its preference list and an adjacency list of the directed graph.
+        graph[i].resize(n);
```

Comment:
use tabs (please, _please_ read the style guide when implementing a PR)

Reply:
I'm sorry but I believe I am using tabs? I pasted this code into CLion (which should auto format the code) before I pasted it here. Is it faulty or can you point out where I am not using tabs? Thanks @SansPapyrus683

Reply:
fixed